source/ingress.go

```diff
@@ -272,17 +276,27 @@ func endpointsFromIngress(ing *v1beta1.Ingress, ignoreHostnameAnnotation bool, i
 				if host == "" {
 					continue
 				}
-				endpoints = append(endpoints, endpointsForHostname(host, targets, ttl, providerSpecific, setIdentifier)...)
+				definedHostsEndpoints = append(definedHostsEndpoints, endpointsForHostname(host, targets, ttl, providerSpecific, setIdentifier)...)
 			}
 		}
 	}
 
-	// Skip endpoints if we do not want entries from annotations
-	if !ignoreHostnameAnnotation {
-		hostnameList := getHostnamesFromAnnotations(ing.Annotations)
-		for _, hostname := range hostnameList {
-			endpoints = append(endpoints, endpointsForHostname(hostname, targets, ttl, providerSpecific, setIdentifier)...)
-		}
+	// Gather endpoints defined on annotations in the ingress
+	var annotationEndpoints []*endpoint.Endpoint
+	for _, hostname := range getHostnamesFromAnnotations(ing.Annotations) {
+		annotationEndpoints = append(annotationEndpoints, endpointsForHostname(hostname, targets, ttl, providerSpecific, setIdentifier)...)
+	}
+
+	// Determine which hostnames to consider in our final list
+	var endpoints []*endpoint.Endpoint
+	hostnameSourceAnnotation, exists := ing.Annotations[ingressHostnameSourceKey]
+	useDefinedHosts := !exists || strings.ToLower(hostnameSourceAnnotation) != IngressHostnameSourceAnnotationOnlyValue
+	useAnnotationHosts := !exists || strings.ToLower(hostnameSourceAnnotation) != IngressHostnameSourceDefinedHostsOnlyValue
+	if useDefinedHosts {
+		endpoints = append(endpoints, definedHostsEndpoints...)
+	}
+	if !ignoreHostnameAnnotation && useAnnotationHosts {
+		endpoints = append(endpoints, annotationEndpoints...)
```

Comment:
Can we simplify this a bit? It would be better to avoid the vars useDefinedHosts and useAnnotationHosts. Checking also exists before and return earlier would make sense.

Reply:
Agreed - the latest push improves the flow, I think.  Let me know how the latest looks to you, @njuettner.